### ISSUE TITLE
Only display valid sound inputs for Capture Device

### ISSUE THREAD

**Issue by Susurrus** [qTox/qTox#3745]
##### Brief Description

OS: Fedora 24 x64
qTox version: 1.5.1-obs
Commit hash:  f05b7e040a073c7afadaf763dda5beaf130c6dbd
toxcore:  0.0.0
Qt:  5.6.0
Hardware: Dell XPS 13 9434  
…

Configuring my audio input is more difficult than it should be because non-inputs are available in the dropdown.

In my Playback Device settings:
- broadwell-rt286 Analog Stereo
- Built-in Audio Digital Stereo (HDMI)

These both seem correct. But for my audio input it's:
- broadwell-rt286 Analog Stereo
- Monitor of Built-in Audio Digital Stereo (HDMI) 
- Monitor of broadwell-rt286 Analog Stereo

I think the only valid input is the first one, but I'm not certain giving that I'm having microphone problems on this laptop right now that are still being debugged within Fedora. But this gets even more confusing when I add a bluetooth headset (because my microphone doesn't work) when I now have:

In my Playback Device settings:
- broadwell-rt286 Analog Stereo
- Built-in Audio Digital Stereo (HDMI)
- PLT Legend

But for my audio input it's:
- broadwell-rt286 Analog Stereo
- Monitor of Built-in Audio Digital Stereo (HDMI) 
- Monitor of broadwell-rt286 Analog Stereo
- Monitor of PLT Legend
- PLT Legend

This actually led to me filing bug #3733. In this case I want "PLT Legend" selected in both cases, but there's a ton of noise here for extraneous inputs. In Gnome's Settings -> Sound, which is more correct, I see:

Outputs:
- Speakers - broadwell-rt286
- Headset - PLT Legend

Inputs:
- Headset - PLT Legend

This is a much more clear listing of the available sources and sinks on my system. For people who aren't very aware of the various sound devices on their systems, I think this would be a nice little usability win.

**Comment by sudden6** [comment]
Personally I think we should also show the monitor outputs, because this provides an easy way to stream something (requested multiple times IIRC). It may be an idea to hide the monitor outputs behind a checkbox, but I don't think it's a good one.

**Comment by Susurrus** [comment]
Interesting, I hadn't thought about that use case and I definitely think it's valid.

Maybe there's a better way to label things such that it's more clear. I didn't even know what "Monitor of" meant even though I'm a techie, so my less techie friends trying to use this software would likely be even more confused.

Maybe group the outputs with a disabled menu item. So in my previous example you'd instead have:
- Inputs (HEADER, DISABLED)
  - PLT Legend
- Steams (HEADER, DISABLED)
  - Built-in Audio Digital Stereo (HDMI)
  - Broadwell-rt286 Analog Stereo

Though I do not believe the "Monitor of PLT Legend" should show up in the dropdown since it's an input already, and "Monitor of" sounds like it's just for outputs.

**Comment by sudden6** [comment]
>  didn't even know what "Monitor of" meant even though I'm a techie

I don't think we can influence the naming of these, because that's done by Pulseaudio.

Grouping the devices by "Real Hardware" and "Virtual Inputs" or something like that would definitely make sense. I'm just not entirely sure our audio library does provide qTox with the necessary information.

**Comment by zetok** [comment]
![2016-09-25 qtox audio advanced](https://cloud.githubusercontent.com/assets/3148759/18815838/76d92d34-8333-11e6-9592-f4a5a4882980.png)
↑ with `Advanced` ticked on, `Monitor of` and other advanced A/V settings could be shown, otherwise they would be hidden.

By default, `Advanced` should be set to "off".

**Comment by sudden6** [comment]
I don't like to clutter the a/v settings tabs, why not put it in "Advanced"?

**Comment by zetok** [comment]
@sudden6 button? How would one know that there's such button there?

**Comment by sudden6** [comment]
no, an additional checkbox with a text like "Show virtual input devices" in Advanced Settings tab + tooltip and documentation stating what it does.

**Comment by zetok** [comment]
And how one would know that there's setting like that in Advanced tab?

**Comment by sudden6** [comment]
Since this feature is intended for advanced users, I think they will either read the user manual or find it in the Advanced Tab.

**Comment by Susurrus** [comment]
That's why I'm generally not a fan of Advanced tabs. The line between Advanced and Basic is normally fuzzy and it can be frustrating working with a "hidden" set of settings. If there's a setting behind the Advanced tab and I'm explaining to a non-tech-literate friend which setting to change I have to clarify that they need to enable Advanced first, which I prob won't remember if I always have those enabled on my end.

Anyways I think this is a good chance to take a step back and look at how the UI may be refined at a higher level instead of trying to do a small hack here. The main confusion I had in this was understanding the sources and sinks I had at my disposal and how those mapped to my hardware and then also what I wanted to do. I didn't even know cloning an input or output to qtox was possible. This is also how screen sharing is implemented, just a cloned output of your desktop as a virtual hardware device, that this UI can be refactored to make that more explicit.

**Comment by zetok** [comment]
@sudden6 no, it's not really for "advanced users". It's meant to hide unnecessary in 99% of cases complexity. Looking in advanced settings for something that is clearly an **audio** setting is not intuitive and no one would do that.

Also, no one reads the docs/manuals.

@Susurrus 

> (…) This is also how screen sharing is implemented, just a cloned output of your desktop as a virtual hardware device, that this UI can be refactored to make that more explicit. 

Current screen sharing is not how it should be done, and will be done properly one day – there's a separate issue for that.

**Comment by sudden6** [comment]
@zetok 

>  It's meant to hide unnecessary in 99% of cases complexity.

Right, most of our users won't ever need this feature and because of that I think this setting is best placed in the "Advanced" Tab. Which in my opinion is exactly the place where settings that only a small part of our users need should go.

**Comment by zetok** [comment]
> Right, most of our users won't ever need this feature and because of that I think this setting is best placed in the "Advanced" Tab.

Advanced tab is not for hiding settings.

> Which in my opinion is exactly the place where settings that only a small part of our users need should go.

That's simply wrong.

**Comment by zetok** [comment]
Alternatively, to avoid confusion checkbox could be set to `true` by default and have `Simple` label.

**Comment by sudden6** [comment]
> Advanced tab is not for hiding settings.

It's exactly for hiding settings unneeded by the average user.

My point is, adding a checkbox to the a/v settings will not make a complicated settings tab easier. Most of the time things get easier by _removing_ things, not adding additional stuff.

**Comment by zetok** [comment]
> My point is, adding a checkbox to the a/v settings will not make a complicated settings tab easier.

Your point is invalid. The "Advanced" checkbox is something that seems to be widely used e.g. in android ecosystem. Aside from android, you might want to look at how it's done e.g. in chromium. Guess why chrome has >50% of market share?

**It does simplify things for users.**

> Most of the time things get easier by removing things, not adding additional stuff.

And how you want to remove something that shouldn't be removed? Neither it should be in Advanced tab, given that it's an A/V setting.

> > Advanced tab is not for hiding settings.
> 
> It's exactly for hiding settings unneeded by the average user.

Nope. It's for settings that user shouldn't touch, unless they actually know what they're doing, since touching those settings could result in something bad. See the warning on top of the tab.

**Comment by Susurrus** [comment]
I think this issue is getting a little into the weeds as the actual issue is: how do we make it more clear what these options mean to the user.

I really think an easy way to make this more clear is to add categories to the dropdown as I originally proposed in my second comment. Segregate the virtual from the physical inputs and outputs. Before we continue the discussion on the Advanced tab, what do people think about that?

**Comment by zetok** [comment]
> I really think an easy way to make this more clear is to add categories to the dropdown (…)

How exactly it makes things more clear by adding more options that aren't clear at all?

> Segregate the virtual from the physical inputs and outputs.

How exactly would that help someone who has no idea about either? Then there's a question of whether backend lib actually provides that info..

> Before we continue the discussion on the Advanced tab, what do people think about that?

AFAICS provides no benefits. Check, if anything, it makes things worse.

**Comment by ElLamparto** [comment]
There could be another combo "Streaming Source" ("none" by default). If different to "none", during audio call  a switch would appear ("microphone/streaming", for example). So one could talk to a contact, then switch and let him listen to some music, for example. That could be very useful and quite unique in the IM world. Also we would avoid the "Advanced" tab...

**Comment by zetok** [comment]
@ElLamparto that sounds like an option that could be hidden behind `Advanced` switch, otherwise it has the same pitfalls as @Susurrus's idea.

**Comment by ElLamparto** [comment]
@zetok, not quite. If the "Streaming Source" is set to "none" by default, the application will behave like before. And the common user probably will not look there anyway.
